### DISCUSSION TITLE
Dynamic Queries and adding dynamic components' data

### DISCUSSION THREAD

**Discussion by iancormac84** [bevyengine/bevy#11424] [Q&A] [answered]
I'm playing around with the dynamic queries API that @james-j-obrien developed for the game engine, and I'm trying to add String data to my dynamically created components. I have been following the example [here](https://github.com/bevyengine/bevy/blob/main/examples/ecs/dynamic.rs), but I'm having problems adapting it to add a String to the component instead of an array of u64. I'm assuming that is possible, but my programming-fu isn't working. Help? Thanks!

**Comment by iancormac84** [comment]
Just bumping this discussion. My code to try to add String data to the component is below:

`let cell_value = cell.get_value().into_owned(); // A  string

let ptr = unsafe {
&nbsp;&nbsp;&nbsp;&nbsp;let data = std::alloc::alloc_zeroed(info.layout()).cast::&lt;String&gt;();
&nbsp;&nbsp;&nbsp;&nbsp;data.copy_from(&cell_value, size_of::&lt;String&gt;());
&nbsp;&nbsp;&nbsp;&nbsp;let non_null = NonNull::new(data.cast()).unwrap();
&nbsp;&nbsp;&nbsp;&nbsp;OwningPtr::new(non_null)
};`

This compiles, but when the code is run, it eventually exits without running to completion. I know my problem is in the above block of code, but I don't know how to fix it.

**Reply by nicopap** [reply]
What's your goal here? There is a difference between "How do I create a string from raw pointers?" and "How do I use the dynamic component API?"

IMO the example is pretty poor and doesn't describe a realistic use-case. Especially with the custom size `u64` array.

I can see a few errors in the code (notably you are getting a `*mut String`, and copying the `String`'s pointer+len+capacity into an address allocated on the heap rather than copying the actual content of the string, (which is pointed-to by `cell_value.as_ptr()`) But I doubt that's something you want to do in any case. If you were happy with micro-managing memory allocation, I doubt you would choose rust, so I strongly advise you do not use those methods.

In no way do you need `std::alloc` and manually copy data using pointer methods to use the dynamic querying API.

Do you mind if I ask you what is the end goal? I could help create a better examples that could be more useful.

**Reply by iancormac84** [reply]
Thanks for popping in! My understanding of the code in the bevy example I linked is that the component that is created is similar to, for example `struct CompA(&[u64]);`.

Assuming that my understanding is correct, I want to create similar to this, except that the component would contain a String. Thanks.

**Reply by james-j-obrien** [reply, marked as answer]
You have a pretty good understanding but there is one wrinkle in that the type in the example actually represents `struct CompA([u64, X])` where X is some dynamically defined size (note the lack of a reference, it's stored inline). The purpose in using that type for the example is to show that the dynamic components can have their sizes chosen dynamically as well, for use cases where you don't know what data you need to store. I do understand nicopap's argument that it is somewhat niche however this is the only way to demonstrate "fully" dynamic components, we could also add a simpler variant that uses a known layout. 

Since you know the exact layout of the data you want to store in your components your use case is simpler and you don't have to manually allocate the memory with `alloc` (I didn't have to when writing the example either, I've made a PR to fix it). Instead of creating the `OwningPtr` from a raw pointer you can instead do this:
```rust
// If you can consume the ptr inline
OwningPtr::make("hello".to_string(), |ptr| {
  // Do something with your ptr here
});

// If you need to store it to insert later
let mut string = "hello".to_string();
// Cast to `u8` is required as `OwningPtr` only accepts `NonNull<u8>`
let non_null = NonNull::from(&mut string).cast::<u8>();
// Note this is unsafe as you need to ensure the pointed to data is not free'd before it's copied into the components
let ptr = unsafe { OwningPtr::<Aligned>::new(non_null) };
```

There should definitely be a better way to do this that doesn't require so much wrangling, but if there is I'm not aware of it.

As nicopap pointed out the copy you were doing is from a `String` to a `*mut String` causing undefined behaviour.

**Reply by iancormac84** [reply]
Thank you! I used the `OwningPtr::make` option and used `EntityWorldMut::insert_by_id` to insert the data.
